### PR TITLE
Disallow module attribute mutation

### PR DESCRIPTION
**Pull request by mergennachin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88354

Summary:

See https://github.com/pytorch/torchdynamo/issues/1475

Not allowing any new mutations happen inside forward() function during
export.

Test Plan:

Run `python test/dynamo/test_export.py` and make sure it passes

Added new unit tests (3 positive tests and 4 negative tests)

Here's what the actual error looks like

```
  File "/home/mnachin/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 322, in step
    getattr(self, inst.opname)(inst)
  File "/home/mnachin/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 835, in STORE_ATTR
    assert not self.export, f"Mutating module attribute {inst.argval} during export."
AssertionError: Mutating module attribute a during export.

from user code:
   File "/data/users/mnachin/pytorch/test/dynamo/test_export_mutations.py", line 25, in forward
    self.a = self.a.to(torch.float64)

Set torch._dynamo.config.verbose=True for more information
```  

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

**Review by tugsbayasgalan — COMMENTED** [review]
Can you add testcases for dynamo non-export mode? So that we can verify that it only fails for export mode

File: test/dynamo/test_export_mutations.py
Comment (tugsbayasgalan) [inline test/dynamo/test_export_mutations.py:93]: Can you add a check that the output of this gm is consistent with eager mode.

**Review by tugsbayasgalan — APPROVED** [review]
looks good to me. But let's wait for dynamo team's approval

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**
